appserver/handle_topology.go

```diff
@@ -0,0 +1,531 @@
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	return ws
+}
+
+// Create a watcher for Deployments and DeploymentConfigs.
+func createNodeWatcher(namespace string, k *kubeclient.KubeClient) *watcher.Watch {
+	listOptions := metav1.ListOptions{}
+	onGetWatchError := func(err error) {
+		fmt.Errorf("Error is %+v", err)
+	}
+	newWatch := watcher.NewWatch(namespace,
+		k,
+		k.GetDeploymentConfigWatcher(namespace, listOptions, onGetWatchError),
+		k.GetDeploymentWatcher(namespace, listOptions, onGetWatchError),
+	)
+	newWatch.SetFilters([]watch.EventType{watch.Added, watch.Modified, watch.Deleted})
+	newWatch.StartWatcher()
+
+	return newWatch
+}
+
+// Create a watcher for all resources.
+func createResourceWatcher(namespace string, opts metav1.ListOptions, k *kubeclient.KubeClient) *watcher.Watch {
+	onGetWatchError := func(err error) {
+		fmt.Errorf("Error is %+v", err)
+	}
+	newWatch := watcher.NewWatch(namespace,
+		k,
+		k.GetDeploymentConfigWatcher(namespace, opts, onGetWatchError),
+		k.GetDeploymentWatcher(namespace, opts, onGetWatchError),
+		k.GetReplicationControllerWatcher(namespace, opts, onGetWatchError),
+		k.GetReplicaSetWatcher(namespace, opts, onGetWatchError),
+		k.GetServiceWatcher(namespace, opts, onGetWatchError),
+		k.GetRouteWatcher(namespace, opts, onGetWatchError),
+	)
+	newWatch.SetFilters([]watch.EventType{watch.Added, watch.Modified, watch.Deleted})
+	newWatch.StartWatcher()
+
+	return newWatch
+}
+
+// Compile the metav1.ListOptions for resources.
+func getResourcesListOptions(nMetaMap map[string][]nodeMeta) map[metav1.ListOptions]nodeMeta {
+	listOptions := make(map[metav1.ListOptions]nodeMeta)
+
+	// For each node, create the metav1.ListOptions based off
+	// the app.kubernetes.io/name label.
+	for lKey, nMetas := range nMetaMap {
+		if lKey != "" {
+			options := metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app.kubernetes.io/name=%s", lKey),
+			}
+			for _, meta := range nMetas {
+				listOptions[options] = meta
+			}
+		}
+	}
+
+	return listOptions
+}
+
+// Gets node metadata.
+func getNodeMetadata(event watch.Event) nodeMeta {
+	var x interface{} = event.Object
+	var node nodeMeta
+	switch x.(type) {
+	case *deploymentconfigv1.DeploymentConfig:
+		dc := x.(*deploymentconfigv1.DeploymentConfig)
+		node = nodeMeta{
+			ID:          base64.StdEncoding.EncodeToString([]byte(dc.UID)),
+			Name:        dc.Name,
+			Kind:        "DeploymentConfig",
+			Type:        "workload",
+			Value:       x,
+			Labels:      dc.Labels,
+			Annotations: dc.Annotations,
+		}
+	case *appsv1.Deployment:
+		d := x.(*appsv1.Deployment)
+		node = nodeMeta{
+			ID:          base64.StdEncoding.EncodeToString([]byte(d.UID)),
+			Name:        d.Name,
+			Kind:        "Deployment",
+			Type:        "workload",
+			Value:       x,
+			Labels:      d.Labels,
+			Annotations: d.Annotations,
+		}
+	default:
+		fmt.Println(reflect.TypeOf(x))
```

Comment:
The default case should be handled properly instead of `fmt.Println`. @tinakurian ?

Reply:
It should be an error because the watcher is assigned to watch only these two types. Will update this!

Reply:
updated here: 2d6b95c5b7ff415af6ac63420f1487983bbd45a5